File: backend/utils/translation_filter.py
import re

class TranslationFilter:    
    # Patterns that should NOT be translated
    PATTERNS = {
        'email': r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b',
        'phone': r'[\+]?[(]?[0-9]{3}[)]?[-\s\.]?[0-9]{3}[-\s\.]?[0-9]{4,6}\b',
        'url': r'https?://[^\s]+|www\.[^\s]+',
        'pure_number': r'^[\d,]+(\.\d+)?$',  # Only numbers (with optional commas)
        'id': r'(?i)^(?:#\d+|ID:\s*\d+|id=\d+)$',
        'html_tag': r'<[^>]+>',  # HTML tags
        'special_char_only': r'^[^\w\s]+$',  # Only special characters
        'currency_code': r'^\s*(USD|EUR|GBP|INR|AED|CAD|AUD|SGD|JPY|CNY|KRW|BRL|ZAR|TRY|MYR|THB|IDR|PHP|CZK|PLN|DKK|SEK|NOK|HUF|ILS|VND|NGN|EGP|MXN)\s*$',
        'currency_symbol': r'^\s*[\$€£¥₹₩₺₪฿₫]\s*$',
        'price_amount': r'^(?:[\$€£¥₹₩₺₪฿₫]|Rs\.?|Ksh|R\$|kr\.?)?\s*[\d,]+\.?\d*\s*[-–]\s*(?:[\$€£¥₹₩₺₪฿₫]|Rs\.?|Ksh|R\$|kr\.?)?\s*[\d,]+\.?\d*$',  # Price ranges
        'price_with_symbol': r'^(?:[\$€£¥₹₩₺₪฿₫]|Rs\.?|Ksh|R\$|kr\.?)\s*[\d,]+\.?\d*$',  # Symbol prefixed: $19.99, Rs. 100
        'price_with_suffix': r'^[\d,]+\.?\d*\s*(?:[\$€£¥₹₩₺₪฿₫]|Rs\.?|Ksh|R\$|kr\.?|€)$',  # Symbol suffixed: 19.99€
        'price_with_code': r'^[\d,]+\.?\d*\s+(USD|EUR|GBP|INR|AED|CAD|AUD|SGD|JPY|CNY|KRW|BRL|ZAR|TRY|MYR|THB|IDR|PHP|CZK|PLN|DKK|SEK|NOK|HUF|ILS|VND|NGN|EGP|MXN)$',
        'code_with_price': r'^(USD|EUR|GBP|INR|AED|CAD|AUD|SGD|JPY|CNY|KRW|BRL|ZAR|TRY|MYR|THB|IDR|PHP|CZK|PLN|DKK|SEK|NOK|HUF|ILS|VND|NGN|EGP|MXN)\s*[\d,]+\.?\d*$',
        'from_price': r'(?i)^(?:From|Starting at|As low as)\s+(?:[\$€£¥₹₩₺₪฿₫]|Rs\.?|Ksh|R\$|kr\.?)?\s*[\d,]+\.?\d*$',
        'price_with_label': r'(?i)^(?:Regular price|Sale price|Unit price|Price)\s*:?\s*(?:(?:[\$€£¥₹₩₺₪฿₫]|Rs\.?|Ksh|R\$|kr\.?)\s*[\d,]+\.?\d*|[\d,]+\.?\d*\s*(?:USD|EUR|GBP|INR|AED|CAD|AUD|SGD|JPY|CNY|KRW|BRL|ZAR|TRY|MYR|THB|IDR|PHP|CZK|PLN|DKK|SEK|NOK|HUF|ILS|VND|NGN|EGP|MXN))$',
        'price_with_label_suffix': r'(?i)^(?:(?:[\$€£¥₹₩₺₪฿₫]|Rs\.?|Ksh|R\$|kr\.?)\s*[\d,]+\.?\d*|[\d,]+\.?\d*\s*(?:USD|EUR|GBP|INR|AED|CAD|AUD|SGD|JPY|CNY|KRW|BRL|ZAR|TRY|MYR|THB|IDR|PHP|CZK|PLN|DKK|SEK|NOK|HUF|ILS|VND|NGN|EGP|MXN))\s*(?:Sale|Regular price)$',
        'price_words_and_amount': r'(?i)^(?:Regular price|Sale price)?\s*(?:[\$€£¥₹₩₺₪฿₫]|Rs\.?|Ksh|R\$|kr\.?)\s*[\d,]+\.?\d*\s*(?:Sale|Regular price)?$',
        'price_with_currency_code_suffix': r'^(?:[\$€£¥₹₩₺₪฿₫]|Rs\.?|Ksh|R\$|kr\.?)\s*[\d,]+\.?\d*\s+(USD|EUR|GBP|INR|AED|CAD|AUD|SGD|JPY|CNY|KRW|BRL|ZAR|TRY|MYR|THB|IDR|PHP|CZK|PLN|DKK|SEK|NOK|HUF|ILS|VND|NGN|EGP|MXN)$',
    }
    
    @staticmethod
    def should_skip(text):
        """
        Determine if text should be skipped from translation.
        Returns True if text matches any exclusion pattern.
        """
        if not text or not isinstance(text, str):
            return False
        
        text = text.strip()
        
        # Check each pattern
        for pattern_name, pattern in TranslationFilter.PATTERNS.items():
            if re.search(pattern, text):
                return True
        
        return False
    
    @staticmethod
    def filter_translatable_content(texts):
        """
        Filter a list of texts and return only translatable ones.
        Returns dict with indices and their corresponding texts.
        
        Args:
            texts: List of strings to filter
            
        Returns:
            dict: {'translatable_indices': [indices], 'texts': [texts], 'skipped': {'index': 'reason'}}
        """
        if not isinstance(texts, list):
            texts = [texts]
        
        translatable = []
        skipped = {}
        
        for idx, text in enumerate(texts):
            if TranslationFilter.should_skip(text):
                skipped[idx] = text
            else:
                translatable.append({'original_idx': idx, 'text': text})
        
        return {
            'translatable': translatable,
            'skipped': skipped,
            'total': len(texts),
            'skipped_count': len(skipped)
        }
    
    @staticmethod
    def extract_safe_content(text):
        """
        Extract safe translatable content while preserving non-translatable parts.
        Splits text into translatable and non-translatable chunks.
        
        Args:
            text: String to process
            
        Returns:
            List of dicts with 'type' ('translatable'/'skip') and 'content'
        """
        if not text:
            return []
        
        if TranslationFilter.should_skip(text):
            return [{'type': 'skip', 'content': text}]
        
        result = []
        current_chunk = ""
        i = 0
        
        while i < len(text):
            # Try to match email
            email_match = re.search(TranslationFilter.PATTERNS['email'], text[i:])
            if email_match and email_match.start() == 0:
                if current_chunk:
                    result.append({'type': 'translatable', 'content': current_chunk})
                    current_chunk = ""
                result.append({'type': 'skip', 'content': email_match.group()})
                i += email_match.end()
                continue
            
            # Try to match phone
            phone_match = re.search(TranslationFilter.PATTERNS['phone'], text[i:])
            if phone_match and phone_match.start() == 0:
                if current_chunk:
                    result.append({'type': 'translatable', 'content': current_chunk})
                    current_chunk = ""
                result.append({'type': 'skip', 'content': phone_match.group()})
                i += phone_match.end()
                continue
            
            # Try to match URL
            url_match = re.search(TranslationFilter.PATTERNS['url'], text[i:])
            if url_match and url_match.start() == 0:
                if current_chunk:
                    result.append({'type': 'translatable', 'content': current_chunk})
                    current_chunk = ""
                result.append({'type': 'skip', 'content': url_match.group()})
                i += url_match.end()
                continue
            
            current_chunk += text[i]
            i += 1
        
        if current_chunk:
            result.append({'type': 'translatable', 'content': current_chunk})
        
        return result

File: frontend/src/utils/translationFilter.js
export const TranslationFilterUtils = {
  patterns: {
    email: /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g,
    phone: /[\+]?[(]?[0-9]{3}[)]?[-\s\.]?[0-9]{3}[-\s\.]?[0-9]{4,6}\b/g,
    url: /https?:\/\/[^\s]+|www\.[^\s]+/g,
    pureNumber: /^\d+(\.\d+)?$/,
    id: /#\d+|ID:\s*\d+|id=\d+/g,
    htmlTag: /<[^>]+>/g,
    specialCharOnly: /^[^\w\s]+$/,
  },

  /**
   * Check if a text should be skipped from translation
   * @param {string} text - Text to check
   * @returns {boolean} - True if text should be skipped
   */
  shouldSkip(text) {
    if (!text || typeof text !== 'string') return false;

    text = text.trim();

    // Check email
    if (this.patterns.email.test(text)) {
      this.patterns.email.lastIndex = 0; // Reset regex state
      return true;
    }

    // Check phone
    if (this.patterns.phone.test(text)) {
      this.patterns.phone.lastIndex = 0;
      return true;
    }

    // Check URL
    if (this.patterns.url.test(text)) {
      this.patterns.url.lastIndex = 0;
      return true;
    }

    // Check pure number
    if (this.patterns.pureNumber.test(text)) return true;

    // Check ID pattern
    if (this.patterns.id.test(text)) {
      this.patterns.id.lastIndex = 0;
      return true;
    }

    // Check HTML tag
    if (this.patterns.htmlTag.test(text)) {
      this.patterns.htmlTag.lastIndex = 0;
      return true;
    }

    // Check special characters only
    if (this.patterns.specialCharOnly.test(text)) return true;

    return false;
  },

  /**
   * Filter an array of texts and return only translatable ones
   * @param {Array<string>} texts - Array of texts to filter
   * @returns {Object} - {translatable: Array, skipped: Object, count: Object}
   */
  filterTranslatableContent(texts) {
    if (!Array.isArray(texts)) texts = [texts];

    const translatable = [];
    const skipped = {};

    texts.forEach((text, idx) => {
      if (this.shouldSkip(text)) {
        skipped[idx] = text;
      } else {
        translatable.push({ originalIdx: idx, text });
      }
    });

    return {
      translatable,
      skipped,
      total: texts.length,
      skippedCount: Object.keys(skipped).length,
    };
  },

  /**
   * Extract translatable content while preserving non-translatable parts
   * @param {string} text - Text to process
   * @returns {Array} - Array of {type: 'translatable'|'skip', content: string}
   */
  extractSafeContent(text) {
    if (!text) return [];

    if (this.shouldSkip(text)) {
      return [{ type: 'skip', content: text }];
    }

    const result = [];
    let currentChunk = '';
    let i = 0;

    while (i < text.length) {
      // Try to match email
      const emailMatch = text.substring(i).match(this.patterns.email);
      if (emailMatch && emailMatch.index === 0) {
        if (currentChunk) {
          result.push({ type: 'translatable', content: currentChunk });
          currentChunk = '';
        }
        result.push({ type: 'skip', content: emailMatch[0] });
        i += emailMatch[0].length;
        this.patterns.email.lastIndex = 0;
        continue;
      }

      // Try to match phone
      const phoneMatch = text.substring(i).match(this.patterns.phone);
      if (phoneMatch && phoneMatch.index === 0) {
        if (currentChunk) {
          result.push({ type: 'translatable', content: currentChunk });
          currentChunk = '';
        }
        result.push({ type: 'skip', content: phoneMatch[0] });
        i += phoneMatch[0].length;
        this.patterns.phone.lastIndex = 0;
        continue;
      }

      // Try to match URL
      const urlMatch = text.substring(i).match(this.patterns.url);
      if (urlMatch && urlMatch.index === 0) {
        if (currentChunk) {
          result.push({ type: 'translatable', content: currentChunk });
          currentChunk = '';
        }
        result.push({ type: 'skip', content: urlMatch[0] });
        i += urlMatch[0].length;
        this.patterns.url.lastIndex = 0;
        continue;
      }

      currentChunk += text[i];
      i += 1;
    }

    if (currentChunk) {
      result.push({ type: 'translatable', content: currentChunk });
    }

    return result;
  },

  /**
   * Reconstruct text from extracted content chunks
   * @param {Array} chunks - Array of {type, content}
   * @param {string} translatedText - Translated content (for translatable parts)
   * @returns {string} - Reconstructed text
   */
  reconstructContent(chunks, translatedText) {
    const translatableChunks = chunks
      .filter(c => c.type === 'translatable')
      .map(c => c.content);

    if (translatableChunks.length === 0) return translatedText;

    const translatableParts = translatedText.split('\n');
    let resultIndex = 0;

    return chunks
      .map(chunk => {
        if (chunk.type === 'skip') return chunk.content;
        return translatableParts[resultIndex++] || chunk.content;
      })
      .join('');
  },
};

export default TranslationFilterUtils;
